### DISCUSSION TITLE
Drawings extraction from PDF

### DISCUSSION THREAD

**Discussion by meghanaviyyapu** [pymupdf/PyMuPDF#1577] [Looking for help]
I would like to extract the rect values of specific drawings from PDF.
For example, in the following PDF, I just want to extract rect values of the tick marks present on the first page. Can you please let me know how this can be achieved.
[EPack1.pdf](https://github.com/pymupdf/PyMuPDF/files/7983463/EPack1.pdf)

I have tried extracting drawings using the following code I have got all the drawings present on that page. Attaching the output PDF.
[drawings-page-0.pdf](https://github.com/pymupdf/PyMuPDF/files/7983470/drawings-page-0.pdf)

page = doc[0]
paths = page.get_drawings()
for path in paths:
# ------------------------------------
# draw each entry of the 'items' list
# ------------------------------------
    for item in path["items"]: # these are the draw commands
        if item[0] == "l": # line
            shape.draw_line(item[1], item[2])
        elif item[0] == "re": # rectangle
            shape.draw_rect(item[1])
        elif item[0] == "qu": # quad
            shape.draw_quad(item[1])
        elif item[0] == "c": # curve
            shape.draw_bezier(item[1], item[2], item[3], item[4])
        else:
            raise ValueError("unhandled drawing", item)
# ------------------------------------------------------
# all items are drawn, now apply the common properties
# to finish the path
# ------------------------------------------------------
        shape.finish(
        fill=path["fill"], # fill color
        color=path["color"], # line color
        dashes=path["dashes"], # line dashing
        even_odd=path.get("even_odd", True), # control color of overlaps
        closePath=path["closePath"], # whether to connect last and first point
        lineJoin=path["lineJoin"], # how line joins should look like
        lineCap=max(path["lineCap"]), # how line ends should look like
        width=path["width"], # line width
        stroke_opacity=path.get("stroke_opacity", 1), # same value for both
        fill_opacity=path.get("fill_opacity", 1),) # opacity parameters
shape.commit()
outpdf.save("drawings-page-0.pdf")

**Comment by JorjMcKie** [comment]
There is no general recipe for doing this.
You need to put in more information you may have from somewhere else: in your case you might want to only look for drawings that contain curves ("c" items), or lines that are nor parallel to axes.

**Reply by meghanaviyyapu** [reply]
If I have drawings content, can I extract the rect value of drawing content from the below data
![image](https://user-images.githubusercontent.com/36301306/152126435-2bdf77df-2ae9-43bd-8230-a030aac9868f.png)

**Reply by JorjMcKie** [reply]
Well, these data are exactly what is being used to create the output of `Page.get_drawings()`. So if you try to do this yourself, you will probably be not be as good by far ... 😉

**Reply by JorjMcKie** [reply]
MuPDF's cli `mutool` has a trace feature that produces XML-like output to inspect a page's painting actions: `mutool trace EPack1.pdf`.
Compared to the raw PDF source you printed out, the tool's advantage is that all the coordinate computations have already been cleaned up. And of course the XML syntax is a help.

**Reply by meghanaviyyapu** [reply]
Thanks for your reply.  My objective is to highlight specific drawings after extracting the specific drawing content from the content stream. Can you help me with any snippets for achieving this?

**Reply by JorjMcKie** [reply]
There is nothing special required here. Once you have identified the drawing of interest (i.e. you know its rect), simply add a highlight annotation.
In contrast to a common misconception, highlight annotations (as well as underline, strikeout, ...) do not require any text - just providing a rectangle is sufficient.

**Reply by meghanaviyyapu** [reply]
I have identified the required drawing data in the content stream but I don't have rect value of this data. Can you please let me know how I can get rect value using this data?
![image](https://user-images.githubusercontent.com/36301306/152136067-c58c7bd5-df16-4cd3-b8ff-1cd44929e7b8.png)

**Comment by JorjMcKie** [comment]
As I wrote: it does not make sense to back-compute coordinates starting from the `/Contents`. But if you absolutely insist, you should look up the information in the PDF spec  "Annex A - Operator Summary", pages 643 and following.
If you manage to locate this you should instead use the drawings output and identify the object of interest there.
Or use the `mutool trace` output - it will at least spare you the dozens of complications that may occur when computing the correct coordinates

**Comment by meghanaviyyapu** [comment]
I have tried extracting all the rect values of drawings present on one page and identified one drawing of interest in the stream but could not identify any re value of the specific drawing in the stream. Can you let me know how page.get_drawings() gives rect values when the drawing in the content stream does not have re operator.

Code:
![image](https://user-images.githubusercontent.com/36301306/152297425-5316f490-4699-4e58-a5b0-dd71e1ee6fce.png)
Identified Object of Interest Data:
![image](https://user-images.githubusercontent.com/36301306/152297490-a6c392ad-f828-479d-8f92-7efcf62fbe8f.png)

**Comment by JorjMcKie** [comment]
The `path["rect"]` value is computed from the union of all points of all `path["items"]`, so it is the smallest rectangle covering all lines, curves, rectangles and quads.

**Reply by meghanaviyyapu** [reply]
Can you let me know how headers and footers can be extracted from a PDF?

**Reply by JorjMcKie** [reply]
This is just normal text. Finding out whether or not something is a header / footer is up to you.

**Comment by JorjMcKie** [comment]
> When I tried highlighting the drawing using the BBox values, the rectangle was getting drawn on different part of the page. Can you help me with this.

Coordinates in the PDF follow the PDF scheme. If you output stuff with PyMuPDF you must use MuPDF's addressing scheme. For translating PDF (e.g. `/BBox`) coordinates use `page.transformation_matrix`.

**Reply by meghanaviyyapu** [reply]
Thanks, now I am able to highlight the drawings.

**Reply by meghanaviyyapu** [reply]
Can you let me know how keys that are set using document.xref_set_key(xref,key,value) can be removed?

**Reply by JorjMcKie** [reply]
> Can you let me know how keys that are set using document.xref_set_key(xref,key,value) can be removed?

Yes, do `doc.xref_set_key(xref, key, "null")`. If the string `key` contains no "/", then a physical removal will take place. Otherwise the value will be set to `null` which plays the role similar to `None` in Python.

**Reply by meghanaviyyapu** [reply]
Thank you

**Reply by meghanaviyyapu** [reply]
Can you let me know how to deal with duplicate symbolic names of images(e.g.one page had two /Im0s ). When I tried extracting the rect values using page.get_image_bbox(symbolic_name), I was getting, "Found multiple images named Im0".

**Reply by JorjMcKie** [reply]
Do not use `get_image_bbox`, but `get_image_rects` instead. This will let you specify the image xref instead of the symbolic name.
But that method will always returna **_list_** of rectangles, because the same image (and hence the same xref for it) may occur multiple times on the same page.

**Reply by JorjMcKie** [reply]
Another thought:
If you do `page.clean_contents()`, then this will rename the reference names of images (and fonts, etc.). After that they should be unique.
A little uncertainty exists in cases, where the page display the same image multiple times ... but this is really rare.

**Reply by meghanaviyyapu** [reply]
Thanks, will try this way

**Comment by JorjMcKie** [comment]
Can you do me - and all the other readers - a favor and start a new discussion for questions that do not fall under the title of this one?
As it is now, looking at the discussion title, nobody will ever suspect all the important things we have been discussing in here.
Thank you!

**Reply by meghanaviyyapu** [reply]
> Can you do me - and all the other readers - a favor and start a new discussion for questions that do not fall under the title of this one? As it is now, looking at the discussion title, nobody will ever suspect all the important things we have been discussing in here. Thank you!


Sure, hereafter I will do that.
